src/khepri_lock.erl

```diff
@@ -0,0 +1,654 @@
+      StoreId :: khepri:store_id(),
+      LockId :: khepri_lock:lock_id(),
+      Scope :: khepri_lock:scope(),
+      Ret :: khepri:ok(Lock) | khepri:error(),
+      Lock :: khepri_lock:lock();
+(LockId, Scope, Options) -> Ret when
+      LockId :: khepri_lock:lock_id(),
+      Scope :: khepri_lock:scope(),
+      Options :: khepri_lock:options(),
+      Ret :: khepri:ok(Lock) | khepri:error(),
+      Lock :: khepri_lock:lock().
+%% @doc Attempts to acquire a lock.
+%%
+%% This function accepts the following two forms:
+%% <ul>
+%% <li>`acquire(StoreId, LockId, Scope)'. Calling it is the same as calling
+%% `acquire(StoreId, LockId, Scope, #{})'.</li>
+%% <li>`acquire(LockId, Scope, Options)'. Calling it is the same as calling
+%% `acquire(StoreId, LockId, Scope, Options)' with the default store ID (see
+%% {@link khepri_cluster:get_default_store_id/0}).</li>
+%% </ul>
+%%
+%% @see acquire/4.
+
+acquire(StoreId, LockId, Scope)
+  when ?IS_STORE_ID(StoreId) andalso ?IS_SCOPE(Scope) ->
+    acquire(StoreId, LockId, Scope, #{});
+acquire(LockId, Scope, Options)
+  when ?IS_SCOPE(Scope) andalso is_map(Options) ->
+    StoreId = khepri_cluster:get_default_store_id(),
+    acquire(StoreId, LockId, Scope, Options).
+
+-spec acquire(StoreId, LockId, Scope, Options) -> Ret when
+      StoreId :: khepri:store_id(),
+      LockId :: khepri_lock:lock_id(),
+      Scope :: khepri_lock:scope(),
+      Options :: khepri_lock:options(),
+      Ret :: khepri:ok(Lock) | khepri:error(),
+      Lock :: khepri_lock:lock().
+%% @doc Attempts to acquire a lock.
+%%
+%% This function attempts to acquire a lock, returning `{ok, Lock}' if it is
+%% successful. If the lock can't be acquired and there are retries remaining,
+%% this function will sleep at random increasing intervals. The first sleep
+%% is at most 1/4 of a second. The maximum value doubles with each failed
+%% attempt to acquire the lock up to a maximum of 8 seconds. The actual time
+%% slept is randomized with {@link random:uniform/1}.
+%%
+%% If the number of retries is exhausted, this function returns `{error,
+%% unavailable}'. The number of retries can be configured with the `retries'
+%% option in {@link options()}. By default, this value is `infinity'
+%% representing unlimited retries.
+%%
+%% This function will throw `{bad_options, Options}' if an invalid combination
+%% of options is passed. For example, setting the `recursive' option to `false'
+%% while providing a `group' is invalid and will cause this function to throw.
+%%
+%% @param StoreId the name of the Khepri store.
+%% @param LockId a term representing the lock.
+%% @param Scope the scope to which the lock should apply.
+%% @param Options options for acquiring the lock.
+%%
+%% @returns `{ok, Lock}' if the lock is successfully acquired or `{error,
+%% unavailable}' if the lock is held and the number of retries are exhausted,
+%% and `{error, Reason}' with any other reason if the acquisition failed for
+%% expected reasons.
```

Comment:
I'm not sure about the return value, especially the `{error, unavailable}`.

To me, failing to acquire a lock because it is held by someone else is not really an error. I was thinking of `{true, Lock}` or `false` instead of `ok` and `{error, unavailable}`, but I'm not sure either.

Reply:
That is closer to what global gives - either `true` for acquired or `false` otherwise. I'll try out `{true, Lock} | false` and we can revisit it if we come up with something better 👍